src/Raven.Server/Smuggler/Documents/MultiShardedDestination.cs

```diff
@@ -242,10 +244,50 @@ public ShardedDocumentActions(ShardedDatabaseContext databaseContext, ByteString
 
             public async ValueTask WriteDocumentAsync(DocumentItem item, SmugglerProgressBase.CountsWithLastEtagAndAttachments progress, Func<ValueTask> beforeFlushing = null)
             {
-                var shardNumber = DatabaseContext.GetShardNumberFor(_allocator, item.Document.Id);
+                var shardNumber = GetShardNumberAndHandleLegacyRevisionsIfNeeded(item.Document);
                 await _actions[shardNumber].WriteDocumentAsync(item, progress, beforeFlushing);
             }
 
+            private int GetShardNumberAndHandleLegacyRevisionsIfNeeded(Document document)
+            {
+                var legacy = document.NonPersistentFlags.Contain(NonPersistentDocumentFlags.LegacyRevision) ||
+                             document.NonPersistentFlags.Contain(NonPersistentDocumentFlags.LegacyHasRevisions);
+                if (legacy == false)
+                    return DatabaseContext.GetShardNumberFor(_allocator, document.Id);
+
+                string id = document.Id;
+                if (document.Data.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata))
```

Comment:
what is wrong with document.TryGetMetadata method? Can't we use it?

Reply:
invert if t reduce nesting

Reply:
> what is wrong with document.TryGetMetadata method? Can't we use it?

we can and will, done :)


> invert if t reduce nesting

done